### PR TITLE
Ensure exported mzML files are valid

### PR DESCRIPTION
**Pull request by jotsetung**
This fixes export of non-valid mzML files (issue #151) and addresses most of the semantic complaints of the OpenMS `FileInfo` tool on mzML files written by `mzR` (see also issue #151).

Validity of mzXML files are not tested, because `FileInfo` uses an older xsd file/format.

Thing that are not or we can not fix:
- Semantic problems picked up from the originating mzML/mzXML file (i.e. `copyWriteMSData` that copies general information from the original file).
- `mzR` has no MS CV term ID (yet), so we get complaints about that.

### CODE REVIEW

File: src/pwiz/data/msdata/IO.cpp [hunk 1]
```diff
@@ -463,15 +463,16 @@ PWIZ_API_DECL void write(minimxml::XMLWriter& writer, const FileDescription& fd)
...
-    attributes.add("count", fd.sourceFilePtrs.size());
-    writer.startElement("sourceFileList", attributes);
-
-    for (vector<SourceFilePtr>::const_iterator it=fd.sourceFilePtrs.begin(); 
-         it!=fd.sourceFilePtrs.end(); ++it)
-         write(writer, **it);
-
-    writer.endElement();
-
+    if (fd.sourceFilePtrs.size() > 0) {
```
Comment (jotsetung) [inline src/pwiz/data/msdata/IO.cpp]: This ensures that no empty `sourceFileList` element is written. An empty `sourceFileList` is not valid according to the mzML standard.

File: src/pwiz/data/msdata/Serializer_mzML.cpp [hunk 2]
```diff
@@ -167,12 +167,19 @@ void Serializer_mzML::Impl::write(ostream& os, const MSData& msd,
     {
         stream_offset indexListOffset = xmlWriter.positionNext();
 
-        XMLWriter::Attributes attributes; 
-        attributes.push_back(make_pair("count", "2"));
+        XMLWriter::Attributes attributes;
```
Comment (jotsetung) [inline src/pwiz/data/msdata/Serializer_mzML.cpp]: These changes ensure that an index element entry is only generated, if also the respective data is present in the file, i.e. a chromatogram index element is only added, if the file contains chromatographic data, a spectrum index element if it contains spectra.

**Comment by jotsetung** [comment]
Sorry, have seen that a lot of changes are due to some automatic re-formatting of the source code (identation etc).

Regarding the changes in proteowizards `IO.cpp` and `Serializer_mzML.cpp` I think it would be best if one of you @sneumann and @lgatto could report this upstream in proteowizard - you are better known in the field.

**Comment by jotsetung** [comment]
Commit https://github.com/sneumann/mzR/pull/153/commits/9a1512c87ec137920f6fa6a4dc118c4eb17b484c includes validation of exported mzML files (using `writeMSData`).

**Comment by jotsetung** [comment]
*I think* it might be save to merge this request - ideally before the next Bioc release - I will also need then some time to fix `MSnbase` to work with the additional `centroided` column returned by `header`. 
@sneumann @lgatto , what you think?

**Comment by jotsetung** [comment]
Thanks. I will then work on the `MSnbase` fix in the meantime.

**Comment by sneumann** [comment]
Hi, yes, Herve Pages had a few commits on .BBSoptions. 
```
commit 8ea1539d919b0215704872a14f58eef14058515d (upstream/master)
Author: Hervé Pagès <hpages@fredhutch.org>
Date:   Mon Mar 12 08:07:20 2018 -0700

    For some unclear reason, using 'make -j8' actually slows down compilation
    of the package on the Windows build machines. Turning off multiple jobs
    compilation for now.

commit 52dd7b66fd23a0dd0e9176f2a08d0240ee7066c0
Author: Hervé Pagès <hpages@fredhutch.org>
Date:   Sat Mar 10 23:25:32 2018 -0800

    Use 8 jobs to compile the package on the Windows build machines.

commit 36104ce857d2dcb2f5752ae299793c6eeec6f1b8
Author: Hervé Pagès <hpages@fredhutch.org>
Date:   Fri Mar 9 18:55:31 2018 -0800

    Oops, removed .gitignore by mistake. Restoring it.

commit 45ac9a15112a13db658865251bee6b21d83e4259
Author: Hervé Pagès <hpages@fredhutch.org>
Date:   Fri Mar 9 18:52:32 2018 -0800

    Trying to speed up installation/compilation of the package on the Windows
    build machines.

commit 953a7c50caa4479110e737afa03bf7f8d07a121e
Author: Hervé Pagès <hpages@fredhutch.org>
Date:   Fri Mar 9 10:13:10 2018 -0800

    Trying to speed up installation/compilation of the package on the build
    machines. Only Linux/Mac for now. Windows will follow (this is actually
    where we're most interested in speeding things up).
```

**Pull request merged**